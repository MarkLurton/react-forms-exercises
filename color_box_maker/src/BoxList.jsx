import React, { useState } from "react";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";
import { v4 as uuid } from "uuid";

const BoxList = () => {
  const INITIAL_STATE = [
    {
      id: uuid(),
      width: 100,
      height: 100,
      color: "red",
    },
    {
      id: uuid(),
      width: 100,
      height: 100,
      color: "cyan",
    },
    {
      id: uuid(),
      width: 100,
      height: 100,
      color: "goldenrod",
    },
  ];
  const [boxes, setBoxes] = useState(INITIAL_STATE);

  const addBox = (box) => {
    let newBox = {
      ...box,
      id: uuid(),
    };
    setBoxes([...boxes, newBox]);
  };

  const removeBox = (id) => {
    setBoxes(boxes.filter((b) => b.id !== id));
  };

  return (
    <div className="BoxList">
      {boxes.map((box) => (
        <Box
          id={box.id}
          key={box.id}
          width={box.width}
          height={box.height}
          color={box.color}
          removeBox={removeBox}
        />
      ))}
      <NewBoxForm addBox={addBox} />
    </div>
  );
};

export default BoxList;
