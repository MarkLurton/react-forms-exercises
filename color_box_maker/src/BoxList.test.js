import React from "react";
import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

it("renders without crashing", function () {
  render(<BoxList />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

it("should add new item", function () {
  const { queryByText, getByLabelText, container } = render(<BoxList />);
  const heightInput = getByLabelText("Height");
  const widthInput = getByLabelText("Width");
  const colorInput = getByLabelText("Color");
  const btn = queryByText("Add Box");
  let boxes = container.getElementsByClassName("Box");
  expect(boxes.length).toBe(3);
  fireEvent.change(heightInput, { target: { value: 100 } });
  fireEvent.change(widthInput, { target: { value: 100 } });
  fireEvent.change(colorInput, { target: { value: "blue" } });
  fireEvent.click(btn);
  boxes = container.getElementsByClassName("Box");
  expect(boxes.length).toBe(4);
});

it("should remove item", function () {
  const { queryAllByText, container } = render(<BoxList />);
  let boxes = container.getElementsByClassName("Box");
  expect(boxes.length).toBe(3);
  const btn = queryAllByText("X");
  fireEvent.click(btn[0]);
  boxes = container.getElementsByClassName("Box");
  expect(boxes.length).toBe(2);
});
