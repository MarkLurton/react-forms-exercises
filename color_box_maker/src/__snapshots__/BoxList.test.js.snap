// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <div
      class="Box"
    >
      <div
        style="width: 100px; height: 100px; background-color: red;"
      />
      <button>
        X
      </button>
    </div>
    <div
      class="Box"
    >
      <div
        style="width: 100px; height: 100px; background-color: cyan;"
      />
      <button>
        X
      </button>
    </div>
    <div
      class="Box"
    >
      <div
        style="width: 100px; height: 100px; background-color: goldenrod;"
      />
      <button>
        X
      </button>
    </div>
    <form>
      <label
        for="height"
      >
        Height
      </label>
      <input
        id="height"
        name="height"
        type="number"
        value="0"
      />
      <label
        for="width"
      >
        Width
      </label>
      <input
        id="width"
        name="width"
        type="number"
        value="0"
      />
      <label
        for="color"
      >
        Color
      </label>
      <input
        id="color"
        name="color"
        type="text"
        value=""
      />
      <button>
        Add Box
      </button>
    </form>
  </div>
</DocumentFragment>
`;
